src/app/api/api.py

```diff
@@ -21,30 +26,87 @@
     field_names_used_for_score_calculation,
 )
 
+
+def get_database(request: Request) -> Database:
+    return request.app.state._db
+
+
 router = APIRouter()
 
+QUALITY_MATRIX_DESCRIPTION = """Calculation of the quality matrix.
+    For each replication source and each property, e.g., `cm:creator`, the quality matrix returns the ratio of
+    elements which miss this entry compared to the total number of entries.
+    A missing entry may be `cm:creator = null`.
+    Additional parameters:
+        store_to_db: Default False. Causes returned quality matrix to also be stored in the backend database."""
+TAG_STATISTICS = "Statistics"
+
 
 @router.get(
     "/quality_matrix",
     status_code=HTTP_200_OK,
     response_model=List[ColumnOutputModel],
     responses={HTTP_404_NOT_FOUND: {"description": "Quality matrix not determinable"}},
-    tags=["Statistics"],
-    description="""Calculation of the quality matrix.
-    For each replication source and each property, e.g., `cm:creator`, the quality matrix returns the ratio of
-    elements which miss this entry compared to the total number of entries.
-    A missing entry may be `cm:creator = null`.""",
+    tags=[TAG_STATISTICS],
+    description=QUALITY_MATRIX_DESCRIPTION,
+)
+async def get_quality_matrix(
+    database: Database = Depends(get_database), store_to_db=False
+):
+    _quality_matrix = await quality_matrix()
+    if store_to_db:
+        await stored_in_timeline(_quality_matrix, database)
+    return _quality_matrix
+
+
+@router.get(
+    "/quality_matrix/{timestamp}",
+    status_code=HTTP_200_OK,
+    response_model=List[ColumnOutputModel],
+    responses={HTTP_404_NOT_FOUND: {"description": "Quality matrix not determinable"}},
+    tags=[TAG_STATISTICS],
+    description=QUALITY_MATRIX_DESCRIPTION
+    + """A timestamp of the format XYZ yields the quality matrix at the respective date.""",
+)
+async def get_past_quality_matrix(
+    timestamp: Optional[int] = None, database: Database = Depends(get_database)
+):
+    if not timestamp:
+        return []
+
+    s = select([Timeline]).where(Timeline.timestamp == timestamp)
+    await database.connect()
+    result: list[Mapping[Timeline]] = await database.fetch_all(s)
```

Comment:
Not sure about the mapping. officially the fetch_all gives you back a `list[Record]` where record in that situation would be a `Timeline` i assume. In general, a `Mapping` type annotation always needs to type parameters: `Mapping[KeyType,ValueType]`

Reply:
Not really - the return type of fetch_all is "typing.List[typing.Mapping]". This function is from databases.
I was playing around with how to map Mapping to Timeline, but failed so far. I can invest some more time, if you wish.

Reply:
Hmm, I was looking at: https://github.com/encode/databases/blob/3fbe52672b95f949ad7603a34dd9c8333f2e1d09/databases/core.py#L145
Maybe thats a newer version?
But it's not really a blocker :)